### ISSUE TITLE
Well-defined behaviour for overlapping classes

### ISSUE THREAD

**Issue by madduck** [rparkins999/CalendarTrigger#18]
I am trying to make it such that my device goes silent on all events by default, but exclude events where I am marked free, or which match Call or Telco or which are all day.

Because the quietest class wins, and I cannot define excluded words, this is not possible.

I think this app would benefit a lot from allowing me to sort event classes and always make the last one win. So did instance, the first class might show a notification and turn the ringer off, while the second one overrides the ringer action with Don'tChange. This would cause only a notification to be shown, as that was overridden, but the ringer would not change at all.

**Comment by rparkins999** [comment]
I usually find it easiest to explicitly tag the events where I want the app to do something: this is the way that the app is designed to be used.

A free/busy test is already supported.

"foo **or** bar" can be achieved by defining two classes with the same actions. Currently I do in fact ignore All Day events, because they don't have meaningful start and end times, and CalendarTrigger takes action at the starts and ends of events. Excluding "Call **or** Telco" is much harder to implement, since it effectively requires a fully general SQL WHERE test. This would require a lot of extra User Interface complexity and also increase the battery drain from the app for all users. Given how aggressive Android has become in trying to stop background services draining the battery, I'm reluctant to go down that road. It also complicates the User Interface a lot.

Incidentally I think that Google are wrong here. If a user wants to run a background service which drains the battery a bit, they should have the choice of doing so. If a rogue app drains the battery a lot without telling the user, most devices will allow the user to find out which app is doing it. In my opinion Google are being overly helpful to device manufacturers who don't put in big enough batteries as opposed to users and app designers who want functionality in the device. CalendarTrigger already has to do a lot of extra work (in fact _increasing_ battery drain) in order to work around Google's limitations.

If enough users reply to this comment wanting a more general SQL test, I'll think about implementing it when I have time. Of course since it is open source you can always fork it and implement it yourself.

Processing classes in order would break the 'quietest wins' rule, which seems to me the right choice for an application whose main function is to stop my phone annoying me (and possibly other people around me) when I'm busy.

**Comment by madduck** [comment]
Hey there, and thanks for your quick response.

You suggest tagging events, in which you'd like your app to do something. I understand that this is how the app was designed.

Unfortunately, I cannot figure out how to make Thunderbird or the Calendar app on Android add words to the description of new events by default.

But if "quietest wins" is the only way to do this, then there will be no way to override a default-silent for events you want to exclude.

Is there another way you can fathom how one could

1. Mute on all events by default
2. Exclude only certain events in certain classes?

Maybe the whole thing could be solved with an implementation of #1?

Where is the logic implemented actually? Is it https://github.com/rparkins999/CalendarTrigger/blob/master/app/src/main/java/uk/co/yahoo/p1rpp/calendartrigger/service/MuteService.java#L1075 ? Because if so, then just removing that `if` would have the effect of the last ringerAction winning, and it should be trivial to add a new configuration option, either per-class that overrides ("force"), or per-app which simply selects "last one wins" over the "most silent one wins". In fact, all four combinations should be easy to implement if it's all done in one place.

Thanks for your consideration!

**Comment by rparkins999** [comment]
No, it isn't as simple as that. The order in which it tests classes would need to be controllable in the UI. This would need an indirection table or inefficient repeated searches through the list of classes. Currently it gives a newly defined class the lowest unused class number, so they are in the order in which you define them unless you have deleted some. So you can't insert a new class in the middle of the sequence without deleting and recreating the classes after it. Also if quietest isn't the default you have to think carefully about what happens if no class matches: at the moment that is easy because it initialises `wantedMode` to no change. As I said in my other reply, getting these things right isn't as simple as it looks. I think a more general SQL search, messy though it is to implement the UI, would be a better solution.

**Comment by madduck** [comment]
I did note that the main screen is empty, and it could list the defined classes with a drag-drop reorder UI. You could initialise the order with the unique IDs by default, but otherwise introduce a new `order` field.

But yes, of course, it's complex.

On the other hand, the abstraction with classes, and the amount of configurability you make available already positions your app in the "much more complex" class than the simple app [Tacere](https://github.com/ciasaboark/Tacere), which unfortunately seems discontinued. Compare https://github.com/ciasaboark/Tacere/issues/52

**Comment by rparkins999** [comment]
Version 3.1 may help you. It has a more general pattern matching algorithm. I will do a release when I've exercised it a lot more on my own phone.

**Comment by madduck** [comment]
Thank you, I'll take a look. Also happy to help you test-drive.

**Comment by rparkins999** [comment]
v3.1 now released with more general pattern matching algorithm.
